### ISSUE TITLE
Proposal: Allow U+0009 in source files

### ISSUE THREAD

**Issue by singpolyma** [purescript/purescript#3939]
## Summary

Purescript is an indentation-based syntax, similar to Haskell.  It would therefore be useful to be able to indent code using the character set aside for this purpose, which is U+0009.

## Motivation

Currently U+0009 is considered a syntax error.

## Proposal

Allow U+0009 and treat it as one level of indentation.  If a layout algorithm that is not aware of indentation is used, treating it as a single whitespace character should suffice.  If a very picky alignment-based algorithm is used, treating it as eight whitespace characters should suffice.

## Examples

```
main = do
	log "hello"
```

**Comment by hdgarrood** [comment]
> Indentation in PureScript is spaces, and I don't think that should change.

Is this more for implementation-based reasons or for stylistic reasons? Do you think the implementation would still be difficult if we made it so that mixing tabs and spaces in a particular source file was disallowed? I was fully on board with disallowing tabs when it happened, but after reading https://www.reddit.com/r/javascript/comments/c8drjo/nobody_talks_about_the_real_reason_to_use_tabs/ I'm less convinced.

**Comment by natefaubion** [comment]
Haskell-style indentation is based on exact columns, it does not have any notion of a semantic tab (unlike Python). Supporting tabs (or any other whitespace that is not a single column) involves making an arbitrary decision about how the compiler translates tabs to spaces. The whitespace rules are already very complicated, and I personally am not keen on extending them at all in any way other than what we already support.

**Comment by natefaubion** [comment]
I should note, that I'm not keen on Haskell-style indentation at all 😆 and vastly prefer more rigid indentation rules. But I just think the way Haskell does it inevitably unifies everyone around spaces.

**Comment by hdgarrood** [comment]
Just to ensure I've understood: the issue of indentation being based on exact columns can manifest in situation where you have code like, for example,

```
let x = f
<tab>y :: T
```
which could parse as
```
let x = f y :: T
```
if a tab counts as 4 columns wide, or as
```
let
  x = f
  y :: T
```
if a tab counts as 8 columns wide?

**Comment by natefaubion** [comment]
Yes.

**Comment by garyb** [comment]
I'd be in favour of this if it preserved the no-mixing-tabs-and-spaces rule, so you could have a tab-only indented file. That does indeed disallow aligned indenting (in the absence of a smart editor), but I think if people really want to use tabs then that's an acceptable tradeoff.

**Comment by singpolyma** [comment]
```
let x = f
<tab>y :: T
```

Would just be abysmally bad style and anyone writing that gets what they deserve.

I think if the layout algorithm is sufficiently close to Haskell that the tab-is-one-whitespace way of seeing it would work well (or, for those who prefer this way of thinking about things: "treat a tab as one space")

**Comment by hdgarrood** [comment]
It's easy to designate that "abysmally bad style" and ignore it, but you can't really justify taking views like that as a programming language designer. The compiler needs to be able to sensibly handle whatever people throw at it, and it especially needs to be able to gracefullly handle things which are close to the boundary of being valid or not. In particular, it makes perfect sense to write
```
let x = f
<tab>y :: T
<tab>y = [...]
```
if you think of a tab as occupying 4 spaces, and I think tab = 4 spaces is the most common default behaviour of terminals and text editors.

**Comment by singpolyma** [comment]
> In particular, it makes perfect sense to write
>```
>let x = f
><tab>y :: T
><tab>y = [...]
>```

That should also be a syntax error in the ideal case. You can't line things 
up using tab characters, only indent them.

I *think* my proposal for one-tab-is-one-space would produce an error for 
this case, would it not?

**Comment by hdgarrood** [comment]
Yes, you'll get "unexpected token 'y'" on the second line. However that's just one example: just because treating a tab as a single whitespace character works okay here, it doesn't mean it will work okay everywhere. For this to be persuasive we really need an argument for why "tab = 1 space" is a sensible choice based on how whitespace works _in general_, not just for a few specific examples.

**Comment by natefaubion** [comment]
It's not clear to me what the benefit of treating it as 1 space gives you. FWIW Haskell treats it as indenting to the next factor of 8. Haskell code indented with tabs is extremely rare.

Sorry, I just don't see why we would want to support something this halfway. I would like to support it equally well as spaces, or not at all, and I don't see how to support it equally well given the current constraints of the rule.

**Comment by rhendric** [comment]
I don't know if this was how pre-0.13 PureScript did things, or if this has been previously considered and rejected, but here's a sketch for how to support tabs in what I think is not a terribly hacky way:

* Tabs are only legal at the beginning of a line. The number of initial tabs in a line is its ‘indent’.
* Replace the column field in `CST.Types.SourcePos` with a pair (indent, column position after tabs).
* Add an indent field to the `Line` constructor of `CST.Types.Comment`. Do some minor futzing around in `CST.Lexer` to make this field non-zero when tabs are found after line breaks.
* Change the ‘delta’ type in `CST.Positions` from a pair (lines, columns) to a sum type (`HDelta` columns|(`VDelta` lines, indent of final line, columns)). `commentDelta` uses the new field of `Line` to set the indent of a `VDelta`, and `applyDelta` uses that to set the indent of a `SourcePos`.
* Change nothing in `CST.Layout`; the natural ordering of pairs is still the ordering on columns that we want when comparing token positions.
* Change `CST.Convert.sourcePos` to add the indent and the column of a CST `SourcePos` to make the column of an AST `SourcePos`, hiding all of this nonsense from any AST consumers. (As far as an IDE or something is concerned, one tab = one character position.)

We could do additional work on top of the above to forbid mixing tab and space indentations, but I don't think the idea of supporting both in this scheme is crazy. Something like this:
```purs
test =
⇥       let x =
⇥       ⇥       [ 0, 1, 2 ]
⇥       ····y = [ 3 ]
⇥       in x <> y
```
would be interpreted correctly, without resorting to any assumptions about how large a tab is.

Have I missed an edge case? Is this a bad idea?

**Comment by natefaubion** [comment]
>  don't know if this was how pre-0.13 PureScript did things

Pre 0.13 just used a fixed column width for tab characters afaik.

**Comment by natefaubion** [comment]
> would be interpreted correctly, without resorting to any assumptions about how large a tab is.
> Have I missed an edge case? Is this a bad idea?

Your let example doesn't quite work because the layout rule is that it must be indented _past_ the column of the first token within the layout block. In the case of `let`, the layout column is determined by the first identifier or pattern. I believe your scheme assumes that the layout rule is determined by the indentation of the keyword that introduces the layout block. It's not clear to me how you assume that two tabs is indented past `x` in the above example.

**Comment by rhendric** [comment]
The ‘column’ of `x` is (1, 5) (that is, one tab, fifth column after the tabs), and the column of the opening `[` is (2, 1). (2, 1) > (1, 5), lexicographically.

**Comment by natefaubion** [comment]
Does that necessitate having a strict tabs-followed-by-spaces indentation? Is tab followed by spaces followed by tabs different tabs followed by tabs followed by spaces of equal values (that is indentation is just reordered)? Do you need to compare the full indentation stack?

I believe on reason Haskell uses multiples of 8 is that you _can_ use tabs in the middle of a line, and the next line can be indented with only tabs, but it does require your code editor/viewer to be smart about tab alignment.
```
example = do<tab>start
                 end
```

**Comment by singpolyma** [comment]
> Does that necessitate having a strict tabs-followed-by-spaces indentation?

Any use of tabs should

**Comment by rhendric** [comment]
I'm proposing only allowing tabs at the beginning of a line, yes. We could always evolve from this scheme to something that keeps a full indentation stack, or to something that allows a tab after non-whitespace and alters the current indentation accordingly. The advantage to my mind of only allowing tabs at the beginning of a line is that it doesn't add much complexity, it lets pure tabbers do their thing, and I think the language thus accepted is likely to be a subset of any reasonable future scheme we would want to implement.

I should probably also say that I'm coming from this from the same perspective as https://github.com/purescript/purescript/issues/3939#issuecomment-697645841: I don't especially care about supporting tabs so that people can do cool alignment tricks, but if there are accessibility concerns that can be mitigated by letting people indent their lines using tabs instead of spaces, I want to make sure we don't artificially restrict who can be productive with PureScript.

**Comment by natefaubion** [comment]
Alright, if we think this can be implemented without severe constraints on formatting style, then I won't stand in the way. I would lprobably ike to make sure that a few larger, "realistic" files (maybe stuff from core like lists) can be formatted and parsed with this convention. If the goal is accessibility, then I'd also like to potentially get input that it is still comfortable to author and consume a file this way. What I'm trying to avoid is going through all the trouble of reintroducing this feature, only to have someone who might want to use it reject it on the grounds that it's still awkward.

**Comment by toastal** [comment]
https://github.com/prettier/prettier/issues/7475

Seems there's a decent chance that Prettier 3.0 will use tabs by default. It will be odd once more code bases are using spaces in PureScript and tabs in the JavaScript FFI (which is what I've been doing, and it works, it's just a bit weird).

**Comment by PureFunctor** [comment]
I've had some time to look into the layout engine and I'm interested in implementing this some time in the near:tm: future, although if anyone else wants to take it up, that'll be fine as well.

**Comment by drathier** [comment]
If the problem with supporting tabs is mixing tabs and spaces, why keep the option of spaces?

**Comment by rhendric** [comment]
I can't imagine how *requiring* every PureScript project to switch to tab-based indentation would be a net positive for the community, if that's what you mean.
